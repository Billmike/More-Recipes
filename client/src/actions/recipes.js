import axios from 'axios';
import {
  ADD_RECIPE,
  GET_RECIPES,
  GET_ONE_RECIPE,
  GET_USER_RECIPES,
  ADD_FAVORITE_RECIPE,
  FETCH_FAVORITE_RECIPES,
  EDIT_RECIPE,
  REMOVE_RECIPE,
  ADD_REVIEW,
  UPVOTE_RECIPE,
  DOWNVOTE_RECIPE
} from './types';
import './toastrConfig';

export const addRecipe = recipe => ({
  type: ADD_RECIPE,
  recipe
});

export const getAllRecipes = (recipes) => {
  return {
    type: GET_RECIPES,
    recipes,
  };
};

export const getOneRecipe = (recipe) => {
  return {
    type: GET_ONE_RECIPE,
    id: recipe.id,
    recipe
  };
};

export const getUserRecipe = (userRecipe) => {
  return {
    type: GET_USER_RECIPES,
    userRecipe,
  };
};

export const addFavorites = (favoriteRecipes) => {
  return {
    type: ADD_FAVORITE_RECIPE,
    favoriteRecipes
  };
};

export const fetchFavorites = (favoriteRecipes) => {
  return {
    type: FETCH_FAVORITE_RECIPES,
    favoriteRecipes,
  };
};

export const editRecipe = (id, updates) => ({
  type: EDIT_RECIPE,
  id,
  updates,
});

export const removeRecipe = ({ id } = {}) => ({
  type: REMOVE_RECIPE,
  id
});

export const addReview = (review) => {
  return {
    type: ADD_REVIEW,
    review,
  };
};

export const upVoteRecipe = (id) => {
  return {
    type: UPVOTE_RECIPE,
    id,
  };
};

export const downVoteRecipe = (id) => {
  return {
    type: DOWNVOTE_RECIPE,
    id,
  };
};

export const startGetAllRecipes = () => {
  return (dispatch) => {
    axios.get('http://localhost:8000/api/v1/recipes')
      .then((res) => {
        dispatch(getAllRecipes(res.data.recipeData));
      })
      .catch((error) => {
        console.log(error);
      });
  };
}

export const startAddRecipe = (recipeData = {}) => {
  return (dispatch) => {
    const {
      name = '',
      description = '',
      img_url = 'no-img-here',
      category = '',
      ingredients = '',
      instructions = ''
    } = recipeData;

    const recipe = {
      name,
      description,
      img_url,
      category,
      ingredients,
      instructions,
    };
    axios.post('http://localhost:8000/api/v1/recipes', recipe)
      .then(() => {
        toastr.success('Recipe added successfully.');
        dispatch(addRecipe({
          id: recipe.id,
          ...recipe
        }));
      });
  };
};

export const startGetUserRecipes = () => {
  return (dispatch) => {
    axios.get('http://localhost:8000/api/v1/users/recipes')
      .then((res) => {
        dispatch(getUserRecipe(res.data.userRecipe));
      })
      .catch((error) => {
        console.log(error);
      });
  };
};

export const startEditRecipe = (id, updates) => {
  return (dispatch) => {
    axios.put(`http://localhost:8000/api/v1/recipes/${id}/modify`, updates)
      .then(() => {
        toastr.success('Recipe edited successfully.');
        dispatch(editRecipe(id, updates));
      })
      .catch((error) => {
        console.log('_<_<_<_<_<__<_', error);
      });
  };
};

export const startRemoveRecipe = (id) => {
  return (dispatch) => {
    axios.delete(`http://localhost:8000/api/v1/recipes/${id}`)
      .then(() => {
        toastr.success('Recipe deleted successfully.');
        dispatch(removeRecipe(id));
      })
      .catch((error) => {
        console.log(error);
      });
  };
};

export const startGetOneRecipe = (id) => {
  return (dispatch) => {
    axios.get(`http://localhost:8000/api/v1/recipes/${id}`)
      .then((res) => {
        dispatch(getOneRecipe(res.data.recipeData));
      })
      .catch((error) => {
        console.log(error);
      });
  };
};

export const startAddFavoriteRecipes = (id) => {
  return (dispatch) => {
    axios.post(`http://localhost:8000/api/v1/recipes/${id}/favorites`)
      .then((res) => {
        toastr.success(res.data.message);
        return dispatch(addFavorites(res.data));
      })
      .catch((err) => {
        if (err.response.data
          .message === 'You need to be logged in to perform this action.') {
          toastr.error(err.response.data.message);
        } else if (err.response.data
          .message === 'Sorry, you cannot perform this action on your own recipe.') {
          toastr.error(err.response.data.message);
        }
      });
  };
};

export const startGetUserFavorites = (id) => {
  return (dispatch) => {
    axios.get(`http://localhost:8000/api/v1/users/${id}/favorites`)
      .then((res) => {
        dispatch(fetchFavorites(res.data));
      })
      .catch((err) => {
        console.log(err);
      });
  };
};

export const startUpvoteRecipe = (id) => {
  return (dispatch) => {
    axios.post(`http://localhost:8000/api/v1/recipes/${id}/votes/upvote`)
      .then((res) => {
        toastr.success(res.data.message);
      })
      .catch((err) => {
        if (err.response.data
          .message === 'You need to be logged in to perform this action.') {
          toastr.error(err.response.data.message);
        } else if (err.response.data
          .message === 'You already upvoted on this recipe') {
          toastr.warning(err.response.data.message);
        }
      });
  };
};

export const startDownVoteRecipe = (id) => {
  return (dispatch) => {
    axios.post(`http://localhost:8000/api/v1/recipes/${id}/votes/downvote`)
      .then((res) => {
        toastr.success(res.data.message);
      })
      .catch((err) => {
        if (err.response.data
          .message === 'You need to be logged in to perform this action.') {
          toastr.error(err.response.data.message);
        } else if (err.response.data
          .message === 'You already downvoted on this recipe') {
          toastr.warning(err.response.data.message);
        }
      });
  };
};

export const startAddReview = (id, data) => {
  return (dispatch) => {
    axios.post(`http://localhost:8000/api/v1/recipes/${id}/reviews`, data)
      .then((res) => {
        dispatch(addReview(res.data));
      })
      .catch((err) => {
        if (err.response.data
          .message === 'You need to be logged in to perform this action.') {
          toastr.error(err.response.data.message);
        } else if (err.response.data
          .message === 'You can\'t post an empty review. Please, enter a happy review for this recipe.') {
          toastr.warning(err.response.data.message);
        }
      });
  };
};
