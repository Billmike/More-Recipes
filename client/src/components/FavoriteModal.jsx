import React from 'react';
import Modal from 'react-modal';

const FavoriteModal = props => (
  <Modal
    isOpen={!!props.selectedFavorite}
    onRequestClose={props.handleClearFavoriteRecipe}
    closeTimeoutMS={300}
    className="modal-class"
    ariaHideApp={false}
  >
    <h3 className="modal-ptag"> Remove recipe</h3>
    <p className="modal-ptag"> Are you sure want to remove this recipe?</p>
    <div className="group-btn">
      <button
        className="btn btn-danger user-btn modal-btn-red"
        onClick={props.removeFavorite}
      >
        Yes
      </button>
      <button
        className="btn user-btn modal-btn-white"
        onClick={props.handleClearFavoriteRecipe}
      >
        No
      </button>
    </div>
  </Modal>
);

export default FavoriteModal;
