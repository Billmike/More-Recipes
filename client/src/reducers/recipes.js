const recipeDefaultState = [];

export default (state = recipeDefaultState, action) => {
  switch (action.type) {
    case 'ADD_RECIPE':
      return [
        ...state,
        action.recipe,
      ];
    default:
      return state;
  }
};
