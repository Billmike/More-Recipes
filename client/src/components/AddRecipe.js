import React from 'react';
import RecipesForm from './RecipesForm';
import { connect } from 'react-redux';
import { addRecipe } from '../actions/recipes';

const AddRecipe = (props) => {
    console.log(props)
	return (
	   <div>
       <h1 className="container">Add Recipe</h1>
        <RecipesForm onSubmit={(recipe) => {
            props.dispatch(addRecipe(recipe));
            props.history.push('/')
        }} />
       </div>	
	)
}

export default connect()(AddRecipe);
