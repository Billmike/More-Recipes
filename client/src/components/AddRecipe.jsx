import React, { Component } from 'react';
import RecipesForm from './RecipesForm';
import { connect } from 'react-redux';
import { addRecipe } from '../actions/recipes';

export class AddRecipe extends Component {
  onSubmit = (recipe) => {
    this.props.addRecipe(recipe);
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <div>
        <h1 className="container"> Add Recipe </h1>
        <RecipesForm
          onSubmit={this.onSubmit}
        />
      </div>
      );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addRecipe: (recipe) => dispatch(addRecipe(recipe))
  }
}

// const AddRecipe = (props) => {
//     console.log(props)
// 	return (
// 	   <div>
//        <h1 className="container">Add Recipe</h1>
//         <RecipesForm onSubmit={(recipe) => {
//             props.dispatch(addRecipe(recipe));
//             props.history.push('/')
//         }} />
//        </div>	
// 	)
// }

export default connect(undefined, mapDispatchToProps)(AddRecipe);
