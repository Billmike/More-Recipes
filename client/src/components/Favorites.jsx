import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Emoji from 'react-emoji-render';
import pizza from '../assets/img/pizzza.jpg';
import { startGetUserFavorites, startAddFavoriteRecipes } from '../actions/recipes';

class Favorites extends Component {
  constructor(props) {
    super(props);
    this.favoriteRecipes = this.favoriteRecipes.bind(this);
  }
  componentDidMount() {
    this.props.startGetUserFavorites(this.props.user.id)
  }

  componentDidUpdate() {
    this.props.startGetUserFavorites(this.props.user.id);
  }

  favoriteRecipes(event) {
    event.preventDefault();
    console.log('Props called here', this.props);
    this.props.startAddFavoriteRecipes(this.props.recipes[0].id);
  }

  render() {
    let favRecipes;
    if (this.props.recipes) {
      favRecipes = this.props.recipes.map((recipe, index) => {
        return (
          <div className="col-md-4">
            <div className="card">
              <Link to={`/recipe/${recipe.id}`}><img className="card-img-top" alt="Pizza" src={ pizza } /></Link>
              <div className="card-body">
                <Link to={`/recipe/${recipe.id}`}><h4 className="card-title">{recipe.name}</h4></Link>
                <p className="card-text">{recipe.description}</p>
                <div className="container action-btn" style={{ textAlign : 'justify' }}>
              <button onClick={this.favoriteRecipes} className="btn btn-danger" role="button">Remove from favs <i className="fa fa-heart"></i></button>
              </div>
              </div>
              <div className="card-footer pad-footer">
                <small className="text-muted small-text">Recipe Category - {recipe.category}</small>
              </div>
            </div>
          </div>
        )
      });
    }
    return (
      <div className="container">
      <h2 className="dashboard-h2"> Welcome to your Dashboard, { this.props.user.username }!</h2>
        <h4 className="dashboard-h4"> My favorites </h4>
        <div className="row">
        { this.props.recipes !== null ? favRecipes : <p className="no-recipes-p"> You currently have no <Emoji text="<3" />. Checkout some recipes <Link to='/'>here.</Link></p> }
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  console.log('Favorites state', state);
  console.log('Favortites props', props);
  return {
    user: state.auth.user,
    recipes: state.recipes.favoriteRecipes
  }
}

export default connect(mapStateToProps, { startGetUserFavorites, startAddFavoriteRecipes })(Favorites);
