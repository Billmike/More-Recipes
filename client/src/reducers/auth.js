import { isEmpty } from 'lodash';
import { SET_CURRENT_USER, GET_USER_INFORMATION } from '../actions/types';

export const initialState = {
  isAuthenticated: !!localStorage.getItem('authToken'),
  userDetails: {},
};

export default (state = initialState, action = {}) => {
  switch (action.type) {
    case SET_CURRENT_USER:
      return {
        ...state,
        isAuthenticated: !isEmpty(action.user),
        userDetails: action.user,
        user: action.user
      };
    case GET_USER_INFORMATION:
      return {
        ...state,
        isAuthenticated: !isEmpty(action.user),
        userDetails: action.user
      };
    default:
      return state;
  }
};
