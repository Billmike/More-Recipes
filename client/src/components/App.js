import React, { Component } from 'react';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import Navbar from './Navbar';
import Homepage from './Hompage';
import Dashboard from './Dashboard';
import Favorite from './Favorite';
import AddRecipe from './AddRecipe';
import EditRecipe from './EditRecipe';
import ProfilePage from './ProfilePage';
import NotFoundPage from './NotFoundPage';

class App extends Component {
    render() {
        return(
            <div>
                <BrowserRouter>
                    <div>
                        <Navbar />
                        <Switch>
                        <Route exact path="/" component={Homepage} />
                        <Route path="/dashboard" component={Dashboard} />
                        <Route path="/favorite" component={Favorite} />
                        <Route path="/add" component={AddRecipe} />
                        <Route path="/edit" component={EditRecipe} />
                        <Route path="/profile" component={ProfilePage} />
                        <Route component={NotFoundPage} />
                    </Switch>
                    </div>
                </BrowserRouter>
            </div>
        )
    }
}

export default App;
