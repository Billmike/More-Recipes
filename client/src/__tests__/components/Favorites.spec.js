import React from 'react';
import { shallow } from 'enzyme';
import { Favorites, mapStateToProps } from '../../components/Favorites';
import FavoriteModal from '../../components/FavoriteModal';
import auth from '../fixtures/authUser';
import { FavoritesRecipeProps } from '../fixtures/recipes';
import state from '../fixtures/state';

const getUserFavoritesAction = jest.fn();
const getUserInformation = jest.fn();

describe('<Favorites />', () => {
  it('Should render the Favorites component', () => {
    const wrapper = shallow(<Favorites
        userDetails={auth}
        getUserInformation={() => Promise.resolve()}
        getUserFavoritesAction={() => Promise.resolve()}
      />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div').length).toBe(3);
  });
  it('Should call the selectFavorite method handler', () => {
    const wrapper = shallow(<Favorites
        recipes={FavoritesRecipeProps}
        getUserInformation={() => Promise.resolve()}
        getUserFavoritesAction={() => Promise.resolve()}
        userDetails={auth}
      />);
    wrapper.find('#selectFavoriteButton').simulate('click');
  });
  it('Should call mapStateToProps', () => {
    mapStateToProps(state);
  });
});
