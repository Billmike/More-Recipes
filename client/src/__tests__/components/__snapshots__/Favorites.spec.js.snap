// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Favorites /> Should render the Favorites component 1`] = `
<div
  id="favorites-container"
>
  <div
    className="container"
  >
    <h2
      className="dashboard-h2"
    >
       
      Welcome to your Dashboard, 
      annabelle
      !
    </h2>
    <h4
      className="dashboard-h4"
    >
       My favorites 
    </h4>
    <div
      className="row"
    >
      <p
        className="no-recipes-p"
      >
         
        You currently have no
        <Emoji
          text="<3"
        />
        . Checkout some recipes
         
        <Link
          id="addFavoritesLink"
          replace={false}
          to="/"
        >
          here.
        </Link>
      </p>
    </div>
  </div>
  <Footer />
</div>
`;
